spiffworkflow-backend/src/spiffworkflow_backend/models/process_instance_file_data.py

```diff
@@ -17,7 +22,40 @@ class ProcessInstanceFileDataModel(SpiffworkflowBaseDBModel):
     mimetype: str = db.Column(db.String(255), nullable=False)
     filename: str = db.Column(db.String(255), nullable=False)
     # this is not deferred because there is no reason to query this model if you do not want the contents
-    contents: str = db.Column(db.LargeBinary().with_variant(LONGBLOB, "mysql"), nullable=False)
+    contents: bytes = db.Column(db.LargeBinary().with_variant(LONGBLOB, "mysql"), nullable=False)
     digest: str = db.Column(db.String(64), nullable=False, index=True)
     updated_at_in_seconds: int = db.Column(db.Integer, nullable=False)
     created_at_in_seconds: int = db.Column(db.Integer, nullable=False)
+
+    def get_contents(self) -> bytes:
+        file_contents = self.contents
+        if current_app.config["SPIFFWORKFLOW_BACKEND_PROCESS_INSTANCE_FILE_DATA_FILESYSTEM_PATH"] is not None:
+            file_contents = self.get_contents_on_file_system()
+        return file_contents
+
+    def store_file_on_file_system(self) -> None:
+        filepath = self.get_full_filepath()
+        os.makedirs(os.path.dirname(filepath))
+        with open(filepath, "w") as f:
+            f.write(self.contents.decode("utf-8"))
+        self.contents = PROCESS_INSTANCE_DATA_FILE_ON_FILE_SYSTEM.encode()
+
+    def get_contents_on_file_system(self) -> bytes:
+        filepath = self.get_full_filepath()
+        with open(filepath) as f:
+            return f.read().encode()
+
+    def get_full_filepath(self) -> str:
+        dir_parts = self.__class__.get_hashed_directory_structure(self.digest)
+        return os.path.join(
+            current_app.config["SPIFFWORKFLOW_BACKEND_PROCESS_INSTANCE_FILE_DATA_FILESYSTEM_PATH"], *dir_parts, self.digest
```

Comment:
Isn't `self.digest` the whole thing? I thought you were trying to reproduce a `Git` type structure where the folder structure is subtracted from the file, so the whole path is the full digest:
```
~/work/spiff-arena deploy-mod-dev
 > ls -l .git/objects/01/fd9e732b52acf6b310e54d1490c8ea1e43f811 
-r--r--r-- 1 jakubgs jakubgs 1388 Feb  7  2023 .git/objects/01/fd9e732b52acf6b310e54d1490c8ea1e43f811

~/work/spiff-arena deploy-mod-dev
 > git cat-file -t 01fd9e732b52acf6b310e54d1490c8ea1e43f811 
blob
```
But it seems like here that would be more like `01/01fd9e732b52acf6b310e54d1490c8ea1e43f811`, which is weird.

Reply:
Yeah, that’s interesting. I went looking for a standard for this and it eluded me. Git’s implementation is definitely an important data point. I like that it is not duplicated in that approach.

I like using the full Digest because you can copy and paste it and slap it into a find –name, so you can very easily restructure the number of prefixed directories, or even attach a file to a GitHub issue without renaming and it remains coherent.

@jbirddog @jasquat ?

Reply:
The point about finding it is a good one, but it seems to me like manual management of those files would be an edge case, more so reserved for developers, less so for administrator and not at all for users. But it makes sense in that regard.

Reply:
My vote is to use the full digest. I don't see a reason to truncate but I can see how the full digest could be helpful. Also it'd be extra logic to rename it.

Reply:
Fair enough. But honestly, what are we trying to optimize with the folder? But I wonder what exactly is the existence of this folder optimizing. In git I think it's simply a matter of limiting number of files per folder.

Reply:
I think limiting number of files in a directory is what we are trying to do as well.